### PR TITLE
perf(semantic): simplify logic in `enter_scope` + `leave_scope`

### PR DESCRIPTION
**Pull request by overlookmotel**
`SemanticBuilder::enter_scope` contained multiple checks purely to handle when called for `Program` scope (where scope has no parent). This is a very uncommon case as `Program` is visited only once, but these checks run for every scope.

So don't call `enter_scope` from `visit_program`, and inline the special logic for root scope there instead, to simplify `enter_scope`.

A branch in `leave_scope` also gets more predictable (always taken).

### CODE REVIEW

**Comment by overlookmotel** [comment]
* **#4383** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4383?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @overlookmotel and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: crates/oxc_semantic/src/builder.rs [hunk 1]
```diff
@@ -460,31 +460,26 @@ impl<'a> SemanticBuilder<'a> {
...
-                // Clone the `CatchClause` bindings and add them to the current scope.
-                // to make it easier to check redeclare errors.
-                let bindings = self.scope.get_bindings(parent_scope_id).clone();
-                self.scope.get_bindings_mut(self.current_scope_id).extend(bindings);
-            }
+        if self.scope.get_flags(parent_scope_id).is_catch_clause() {
+            // Clone the `CatchClause` bindings and add them to the current scope.
+            // to make it easier to check redeclare errors.
+            let bindings = self.scope.get_bindings(parent_scope_id).clone();
+            self.scope.get_bindings_mut(self.current_scope_id).extend(bindings);
```
Comment (Dunqing) [inline crates/oxc_semantic/src/builder.rs:482]: After merging this PR. I'd also like to move this logic to `BlockStatement`

Comment (overlookmotel) [reply crates/oxc_semantic/src/builder.rs:482]: Good idea!

**Comment by overlookmotel** [comment]
I ran the benchmarks several times and perf impact was inconclusive - too much noise in the bench results. But this *should* be a little faster.

**Review by Dunqing — APPROVED** [review]

**Pull request merged**